internal/db/db.go

```diff
@@ -2,281 +2,452 @@
+	deleteCSRStmt  = "DELETE FROM %s WHERE id=$CertificateRequest.id or csr=$CertificateRequest.csr"
 )
 
-// CertificateRequestRepository is the object used to communicate with the established repository.
-type Database struct {
-	certificateTable string
-	usersTable       string
-	conn             *sql.DB
+const (
+	getAllUsersStmt = "SELECT &User.* from %s"
+	getUserStmt     = "SELECT &User.* from %s WHERE id==$User.id or username==$User.username"
+	createUserStmt  = "INSERT INTO %s (username, hashed_password, permissions) VALUES ($User.username, $User.hashed_password, $User.permissions)"
+	updateUserStmt  = "UPDATE %s SET hashed_password=$User.hashed_password WHERE id==$User.id or username==$User.username"
+	deleteUserStmt  = "DELETE FROM %s WHERE id==$User.id"
+	getNumUsersStmt = "SELECT COUNT(*) AS &NumUsers.count FROM %s"
+)
+
+// RetrieveAllCSRs gets every CertificateRequest entry in the table.
+func (db *Database) RetrieveAllCSRs() ([]CertificateRequest, error) {
+	stmt, err := sqlair.Prepare(fmt.Sprintf(getAllCSRsStmt, db.certificateTable), CertificateRequest{})
+	if err != nil {
+		return nil, err
+	}
+	var csrs []CertificateRequest
+	err = db.conn.Query(context.Background(), stmt).GetAll(&csrs)
+	if err != nil {
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return csrs, nil
+		}
+		return nil, err
+	}
+	return csrs, nil
 }
 
-// A CertificateRequest struct represents an entry in the database.
-// The object contains a Certificate Request, its matching Certificate if any, and the row ID.
-type CertificateRequest struct {
-	ID          int
-	CSR         string
-	Certificate string
+// RetrieveCSRbyID gets a CSR row from the repository from a given ID.
+func (db *Database) RetrieveCSRbyID(id int) (*CertificateRequest, error) {
+	csr := CertificateRequest{
+		ID: id,
+	}
+	stmt, err := sqlair.Prepare(fmt.Sprintf(getCSRsStmt, db.certificateTable), CertificateRequest{})
+	if err != nil {
+		return nil, err
+	}
+	err = db.conn.Query(context.Background(), stmt, csr).Get(&csr)
+	if err != nil {
+		return nil, err
+	}
+	return &csr, nil
 }
-type User struct {
-	ID          int
-	Username    string
-	Password    string
-	Permissions int
+
+// RetrieveCSRbyCSR gets a given CSR row from the repository using the CSR text.
+func (db *Database) RetrieveCSRbyCSR(csr string) (*CertificateRequest, error) {
+	row := CertificateRequest{
+		CSR: csr,
+	}
+	stmt, err := sqlair.Prepare(fmt.Sprintf(getCSRsStmt, db.certificateTable), CertificateRequest{})
+	if err != nil {
+		return nil, err
+	}
+	err = db.conn.Query(context.Background(), stmt, row).Get(&row)
+	if err != nil {
+		return nil, err
+	}
+	return &row, nil
 }
 
-var ErrIdNotFound = errors.New("id not found")
+// CreateCSR creates a new CSR entry in the repository. The string must be a valid CSR and unique.
+func (db *Database) CreateCSR(csr string) error {
```

Comment:
```suggestion
func (db *Database) CreateCertificateRequest(csr string) error {
```

Reply:
Generally speaking, I think we abuse the term CSR a bit here and in other places, I made a couple of suggestions to improve this.